packages/codemods/src/codeMods/mods/ComponentToCompat/CompatHelpers.ts

```diff
@@ -0,0 +1,79 @@
+import { SourceFile } from 'ts-morph';
+import { getImportsByPath, appendOrCreateNamedImport, repathImport } from '../../utilities/index';
+
+export interface IComponentToCompat {
+  // Old exact path
+  oldPath: string;
+
+  // The complete path that should either just replace the old one,
+  // Or be where all the compats are imported from.
+  newComponentPath: string;
+
+  // All the components, names, and other things that are exported
+  // from that root folder and that should get repathed
+  namedExports: string[];
+}
+
+export interface IRawCompat {
+  componentName: string;
+  // tslint:disable-next-line: no-any
+  namedExports: { [key: string]: any };
+}
+
+export interface ICompatHash {
+  namedExportsMatch: { [key: string]: string };
+  exactPathMatch: { [key: string]: string };
+}
+
+export function repathNamedImports(file: SourceFile, hash: ICompatHash, indexPath: string) {
+  const imports = getImportsByPath(file, indexPath);
+  imports.forEach(imp => {
+    imp.getNamedImports().forEach(namedImp => {
+      if (hash.namedExportsMatch[namedImp.getName()]) {
+        appendOrCreateNamedImport(file, hash.namedExportsMatch[namedImp.getName()], [namedImp.getStructure()]);
+        namedImp.remove();
+      }
+    });
+
+    // Remove the index import if it no longer has any named imports.
+    if (imp.getNamedImports().length === 0) {
+      imp.remove();
+    }
+  });
+}
+
+// tslint:disable-next-line: no-any
+export function getNamedExports(obj: { [key: string]: any }) {
+  return Object.keys(obj);
+}
+
+export function buildCompatHash(
+  renameCompats: IRawCompat[],
+  getComponentToCompat: (comp: IRawCompat) => IComponentToCompat,
+): ICompatHash {
+  return renameCompats.reduce(
+    (acc: ICompatHash, val) => {
+      const paths = getComponentToCompat(val);
+      acc.exactPathMatch[paths.oldPath] = paths.newComponentPath;
+      paths.namedExports.forEach(path => {
+        acc.namedExportsMatch[path] = paths.newComponentPath;
+      });
+      return acc;
+    },
+    { namedExportsMatch: {}, exactPathMatch: {} },
+  );
+}
+
+export function repathPathedImports(file: SourceFile, hash: ICompatHash) {
+  file.getImportDeclarations().forEach(dec => {
+    const str = dec.getModuleSpecifierValue();
+    if (hash.exactPathMatch[str]) {
```

Comment:
if I'm reading the code right, i don't think this works for the following scenario?

Before:
```
import { DefaultButton, DetailsList } from 'office-ui-fabric-react'
```

After:
```
import { DetailsList } from 'office-ui-fabric-react';
import { DefaultButton } from 'office-ui-fabric-react/lib/compat'`
```

Reply:
Correct! This doesn't, but that's why there is the function `repathNamedImports`. Realistically They should be named better but they do two fundamentally different things. One goes and just checks for an exact match and then replaces that with the new path. That's the one referenced here, `repathPathedImports`. 

The other one finds a specific root path (like office-ui-fabric-react) and checks all of the named imports it contains and then moves those. There's a test for this one too :D

Reply:
Ic now. Looks to me no test is covering this case explicitly. That is to make sure `import { DetailsList } from 'office-ui-fabric-react';` reminds untouched. If i missed it, could you point me to it? If not, could you add the test coverage?

Reply:
I had missed that test! Just added it.

Reply:
I had ones that were close to it, but none that explicitly checked that it left some imports untouched.